File: frontend/src/services/documents.ts
import { api } from "@/lib/api";

export type DocumentStatus = "PENDING" | "PROCESSING" | "READY" | "FAILED";

export interface RecallDocument {
  id: number;
  workspace_id: number;
  original_filename: string;
  file_size: number;
  mime_type: string;
  created_at: string;
  status: DocumentStatus;
  chunk_count: number;
  error_message: string | null;
  updated_at: string | null;
}

export interface DocumentChunk {
  chunk_index: number;
  page_number: number | null;
  content: string;
  token_count: number;
}

export interface StorageUsage {
  used_bytes: number;
  quota_bytes: number;
}

export const documentService = {
  usage() {
    return api.get<StorageUsage>("/documents/usage").then((r) => r.data);
  },

  upload(file: File, opts?: { workspaceId?: number; onProgress?: (percent: number) => void }) {
    const formData = new FormData();
    formData.append("file", file);
    if (opts?.workspaceId != null) formData.append("workspace_id", String(opts.workspaceId));
    return api
      .post<RecallDocument>("/documents/upload", formData, {
        headers: { "Content-Type": "multipart/form-data" },
        onUploadProgress: (evt) => {
          if (opts?.onProgress && evt.total) {
            opts.onProgress(Math.round((evt.loaded / evt.total) * 100));
          }
        },
      })
      .then((r) => r.data);
  },

  list(workspaceId?: number) {
    return api
      .get<RecallDocument[]>("/documents/", {
        params: workspaceId != null ? { workspace_id: workspaceId } : {},
      })
      .then((r) => r.data);
  },

  getChunks(id: number) {
    return api.get<DocumentChunk[]>(`/documents/${id}/chunks`).then((r) => r.data);
  },

  remove(id: number) {
    return api.delete(`/documents/${id}`).then(() => undefined);
  },

  reprocess(id: number) {
    return api.post<RecallDocument>(`/documents/${id}/reprocess`).then((r) => r.data);
  },
};

File: backend/app/api/v1/documents.py
from fastapi import APIRouter, Depends, File, Form, UploadFile, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.activity_event import ActivityEventType
from app.models.user_dtl import User
from app.services.documentService import DocumentService
from app.services.memoryService import MemoryService
from app.services.workspaceService import WorkspaceService

router = APIRouter()


def _document_dict(doc) -> dict:
    return {
        "id": doc.id,
        "workspace_id": doc.workspace_id,
        "original_filename": doc.original_filename,
        "file_size": doc.file_size,
        "mime_type": doc.mime_type,
        "created_at": doc.created_at,
        "status": doc.status.value,
        "chunk_count": doc.chunk_count,
        "error_message": doc.error_message,
        "updated_at": doc.updated_at,
    }


@router.post("/upload", status_code=status.HTTP_202_ACCEPTED)
async def upload_document(
    file: UploadFile = File(...),
    workspace_id: int | None = Form(None),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    workspace = (
        WorkspaceService.get_by_id(db, workspace_id, user.id)
        if workspace_id is not None
        else WorkspaceService.get_or_create_default(db, user.id)
    )
    minio_key, size = await DocumentService.upload_to_minio(file, user.id, db)
    doc = DocumentService.create_record(
        db, user.id, file.filename, minio_key, size, file.content_type, workspace.id
    )
    MemoryService.record(
        db,
        user.id,
        ActivityEventType.DOCUMENT_UPLOADED,
        title=f"Uploaded '{doc.original_filename}'",
        subtitle=f"Queued for processing into {workspace.name}.",
        document_id=doc.id,
    )
    DocumentService.enqueue_processing(doc.id)
    return _document_dict(doc)


@router.get("/")
async def list_documents(
    workspace_id: int | None = None,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    docs = DocumentService.list_for_user(db, user.id, workspace_id)
    return [_document_dict(d) for d in docs]


# Declared before /{document_id} so the literal path wins over the int path param.
@router.get("/usage")
async def get_usage(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return DocumentService.usage(db, user.id)


@router.get("/{document_id}")
async def get_document(
    document_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    doc = DocumentService.get_by_id(db, document_id, user.id)
    MemoryService.track_access(user.id, "document", doc.id)
    return _document_dict(doc)


@router.get("/{document_id}/chunks")
async def get_document_chunks(
    document_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    chunks = DocumentService.get_chunks(db, document_id, user.id)
    return [
        {
            "chunk_index": c.chunk_index,
            "page_number": c.page_number,
            "content": c.content,
            "token_count": c.token_count,
        }
        for c in chunks
    ]


@router.post("/{document_id}/reprocess", status_code=status.HTTP_202_ACCEPTED)
async def reprocess_document(
    document_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    doc = DocumentService.reprocess(db, document_id, user.id)
    return _document_dict(doc)


@router.delete("/{document_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_document(
    document_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    DocumentService.delete(db, document_id, user.id)
